File: c/circularLinkedList.c
#include <stdio.h>
#include <stdlib.h>

// define node structure
struct Node {
  int data;
  struct Node* next;
};

// function to create new node
struct Node* createNode(int data) {
  struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
  newNode->data = data;
  newNode->next = NULL:
  return newNode;
}

// function to insert a node at end 
void insertAtEnd(struct Node** head, int data) {
  struct Node* newNode = createNode(data);
  if(*head == NULL) {
    *head = newNode;
    newNode->next = *head;
  } else {
    struct Node* temp = *head; 
    while(temp->next != *head) {
      temp = temp->next;
    }
    temp->next = newNode;
    newNode->next = *head;
  }
}

// function to print 
void printList(struct Node* head) {
  if(head == NULL) return;
  struct Node* temp = head;
  do {
    printf("%d -> ", temp->data);
    temp = temp->next;
  } while(temp != head);
  printf("(head)\n");
}

int main() {
  struct Node* head = NULL;
  insertAtEnd(&head, 10):
  insertAtEnd(&head, 20);
  insertAtEnd(&head, 30);

  printf("circular linked list; ");
  printList(head);
  return 0;
}

File: python/circularlinked_list.py
class Node:
  def __init__(self, data):
    self.data = data
    self.next = None


class CircularLinkedList:
  def __init__(self):
    self.head = None

  def insert_at_end(self, data):
    new_node = Node(data)
    if not self.head:
      self.head = new_node
      new_node.next = self.head
    else:
      temp = self.head
      while temp.next != self.head:
        temp = temp.next
      temp.next = new_node
      new_node.next = self.head

  def print_list(self):
    if not self.head:
      return
    temp = self.head
    while True:
      print(temp.data, end=" -> ")
      temp = temp.next
      if temp == self.head:
        break
    print("(head)")


# example usage
if __name__ == "__main__":
  cl1 = CircularLinkedList()
  cl1.insert_at_end(10)
  cl1.insert_at_end(20)
  cl1.insert_at_end(30)

  print("circular linked list; ")
  cl1.print_list() 
